platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/results/SerializableTestResultStore.java

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing.results;
+
+import org.apache.commons.io.input.NullReader;
+import org.gradle.api.NonNullApi;
+import org.gradle.api.UncheckedIOException;
+import org.gradle.api.internal.tasks.testing.TestCompleteEvent;
+import org.gradle.api.internal.tasks.testing.TestDescriptorInternal;
+import org.gradle.api.internal.tasks.testing.TestStartEvent;
+import org.gradle.api.tasks.testing.TestMetadataEvent;
+import org.gradle.api.tasks.testing.TestOutputEvent;
+import org.gradle.api.tasks.testing.TestResult;
+import org.gradle.internal.concurrent.CompositeStoppable;
+import org.gradle.internal.serialize.PlaceholderExceptionSupport;
+import org.gradle.internal.serialize.kryo.KryoBackedDecoder;
+import org.gradle.internal.serialize.kryo.KryoBackedEncoder;
+
+import javax.annotation.Nullable;
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+import java.util.zip.ZipOutputStream;
+
+/**
+ * An object that can store test results and their outputs.
+ */
+@NonNullApi
+public final class SerializableTestResultStore {
```

Comment:
AFAIK the original test report code went with a custom format for reasons of CPU and I/O performance, since there is no need for the generic handling of other serialization libraries. But we could go with JSON, I just don't see a major reason to. I also don't want people to mistakenly think the output is stable in any way, although I suppose if it keeps the version it sort of is stable.

Reply:
Yeah, I guess it's a higher-level decision about how we want to do things.  I'd vote that the slight reduction in maintenance effort from not building a custom format is worth the slight performance impact, but it's not my call.

> although I suppose if it keeps the version it sort of is stable.

Right.  And anyone who cares enough to use it can copy-paste our source, so I don't think there's much to this justification.